### PR TITLE
Gradients for Acrobot using AutoDiffScalar + TaylorVar related utility functions in C++

### PR DESCRIPTION
**Pull request by tkoolen**
Added gradients for the AcrobotCPP plant using AutoDiffScalar. The C++ function converts Matlab TaylorVars into Eigen matrices with an AutoDiffScalar<VectorXd> as the scalar type parameter. After calling the dynamics, the results are converted back into TaylorVars. Expanded test to include verification that the gradients are correct.

Also added some utility functions for dealing with TaylorVars and matrices of AutoDiffScalars and cleaned up the way we convert from the various Eigen matrix types back to Matlab a bit by adding a eigenToMatlabGeneral template function with specializations that call eigenToMatlab, eigenToTrigPoly, or eigenToTaylorVar based on the type of the input.

### CODE REVIEW

File: util/drakeMexUtil.h [hunk 1]
```diff
@@ -207,5 +209,57 @@ mxArray* eigenToTrigPoly(const Eigen::Matrix<TrigPolyd,_Rows,_Cols> & trigpoly_m
...
+  mwSize dims[] = {1};
+  prhs[1] = mxCreateCellArray(1, dims);
+  mxArray *plhs[1];
+  mxSetCell(prhs[1], 0, eigenToMatlab(autoDiffToGradientMatrix(m, num_variables)));
+  mexCallMATLABsafe(1, plhs, nrhs, prhs, "TaylorVar");
+  return plhs[0];
+}
+
+template<typename Derived>
+mxArray *eigenToMatlabGeneral(const Eigen::MatrixBase<Derived>& mat)
```
Comment (tkoolen) [inline util/drakeMexUtil.h]: Don't like the name yet. Should we rename eigenToMatlab to eigenToMatlabDouble and call this eigenToMatlab?

File: examples/Acrobot/@AcrobotPlantCpp/manipulatorDynamics.cpp [hunk 2]
```diff
@@ -3,84 +3,88 @@
...
+  C(1) += g * m2 * lc2 * s12;
 
   // damping terms
-  C(0)+=b1*qd(0); C(1)+=b2*qd(1);
+  C(0) += b1 * qd(0);
+  C(1) += b2 * qd(1);
+
+  // input matrix
+  // multiplying by q(0) just to get the size of the derivatives vector right; Matlab TaylorVar operations will complain otherwise. We could handle this case on the Matlab side instead.
+  B << 0.0 * q(0), 1.0;
```
Comment (tkoolen) [inline examples/Acrobot/@AcrobotPlantCpp/manipulatorDynamics.cpp]: The only bit of nastiness. See comment in code. What do you think?

File: util/drakeMexUtil.h (hunk 1 above)
Comment (avalenzu) [inline util/drakeMexUtil.h]: :+1:  for those names.

Comment (tkoolen) [inline util/drakeMexUtil.h]: Let me know in the next day if you have other ideas on this.

**Comment by RussTedrake** [comment]
Code all looks good.  But the build error looks real?

**Comment by tkoolen** [comment]
I'll take a look tomorrow.

**Comment by RussTedrake** [comment]
can you update the git tags for eigen and spotless to reflect our changes today, too?

**Comment by tkoolen** [comment]
OK.

**Comment by tkoolen** [comment]
Found out what was causing the segfaults (unsigned vs. signed integer types bug introduced while fixing warnings on MSVC).

As for the LDLT rankUpdate error (C2244): that seems to be MSVC 2010 having trouble with Eigen code. `template<typename _MatrixType, int _UpLo> class LDLT` contains a function declaration with signature:
```
template <typename Derived>
LDLT& rankUpdate(const MatrixBase<Derived>& w, const RealScalar& alpha=1);
```
where `RealScalar` is typedef'd as `typename NumTraits<typename MatrixType::Scalar>::Real` and `MatrixType` is typedef'd as `_MatrixType`.

`rankUpdate` is implemented out of line with a function signature:
```
template<typename MatrixType, int _UpLo>
template<typename Derived>
LDLT<MatrixType,_UpLo>& LDLT<MatrixType,_UpLo>::rankUpdate(const MatrixBase<Derived>& w, const typename NumTraits<typename MatrixType::Scalar>::Real& sigma)
```
i.e. it directly refers to `NumTraits<typename MatrixType::Scalar>`, whereas the declaration strictly speaking refers to `NumTraits<typename LDLT<MatrixType, _UpLo>::MatrixType::Scalar>`. These are of course the same thing, but MSVC 2010 isn't able to figure that out. After changing the signature in the definition to
```
template<typename MatrixType, int _UpLo>
template<typename Derived>
LDLT<MatrixType,_UpLo>& LDLT<MatrixType,_UpLo>::rankUpdate(const MatrixBase<Derived>& w, const typename NumTraits<typename LDLT<MatrixType,_UpLo>::MatrixType::Scalar>::Real& sigma)
```
I get further in the compilation process, but run into errors involving templates in our own code.

I'm not sure yet what triggered it in this branch; I don't think I added any calls to `lldt` or `solve` or anything like that.

Everything compiled fine in MSVC2015, and I'm hopeful that MSVC2013 would also have no trouble.

**Comment by tkoolen** [comment]
OK, I think I finally got everything. I worked around the rankUpdate error by changing the order of includes in drakeMexUtil.h. I still don't understand what difference this makes, but at least it compiles now on MSVC 2013. I suspect it will also compile on 2010 but haven't tried.

Lots of compiler warnings though, related to multiple assignment operators for Quaternion in 3.2.5. See http://listengine.tuxfamily.org/lists.tuxfamily.org/eigen/2015/06/msg00017.html.

I also had to work around another MSVC issue involving dependent types in template method signatures.

**Comment by tkoolen** [comment]
Finally.

**Comment by tkoolen** [comment]
About the issue with AutoDiffScalar and Polynomial above: here's the Eigen issue I made for it:
http://eigen.tuxfamily.org/bz/show_bug.cgi?id=1057

**Pull request merged**